Boosting without oversampling

In [31]:
import pandas as pd
import catboost
from xgboost import XGBClassifier
from lightgbm import LGBMClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report, confusion_matrix


In [32]:
column_names = ["Suit of card #1", "Rank of card #1", "Suit of card #2", "Rank of card #2", "Suit of card #3", "Rank of card #3", "Suit of card #4", "Rank of card #4", "Suit of card #5", "Rank of card #5", "Poker Hand"]
poker_train = pd.read_csv('poker_train.csv', header=None, names=column_names)
poker_test = pd.read_csv('poker_test.csv', header=None, names=column_names)
print(poker_test.head())
poker_train.head()

   Suit of card #1  Rank of card #1  Suit of card #2  Rank of card #2  \
0                1                1                1               13   
1                3               12                3                2   
2                1                9                4                6   
3                1                4                3               13   
4                3               10                2                7   

   Suit of card #3  Rank of card #3  Suit of card #4  Rank of card #4  \
0                2                4                2                3   
1                3               11                4                5   
2                1                4                3                2   
3                2               13                2                1   
4                1                2                2               11   

   Suit of card #5  Rank of card #5  Poker Hand  
0                1               12           0  
1                2    

Suit of card #1  Rank of card #1  Suit of card #2  Rank of card #2  \
0                1               10                1               11   
1                2               11                2               13   
2                3               12                3               11   
3                4               10                4               11   
4                4                1                4               13   

   Suit of card #3  Rank of card #3  Suit of card #4  Rank of card #4  \
0                1               13                1               12   
1                2               10                2               12   
2                3               13                3               10   
3                4                1                4               13   
4                4               12                4               11   

   Suit of card #5  Rank of card #5  Poker Hand  
0                1                1           9  
1                2                1           9  
2                3                1           9  
3                4               12           9  
4                4               10           9

In [33]:
#train data
poker_ip = poker_train.drop('Poker Hand', axis='columns')
poker_op = poker_train[['Poker Hand']]
#test data
poker_testip = poker_test.drop('Poker Hand', axis='columns')
poker_testop = poker_test[['Poker Hand']]

Cat Boost

In [34]:
cat_features = []
cat_model = catboost.CatBoostClassifier(iterations=100, learning_rate=0.1, depth=6, loss_function='MultiClass')
cat_model.fit (poker_ip, poker_op, cat_features=cat_features)

0:	learn: 1.9996165	total: 18.1ms	remaining: 1.79s
1:	learn: 1.8085867	total: 34.8ms	remaining: 1.7s
2:	learn: 1.6691770	total: 50.6ms	remaining: 1.64s
3:	learn: 1.5608627	total: 65.5ms	remaining: 1.57s
4:	learn: 1.4741428	total: 81.1ms	remaining: 1.54s
5:	learn: 1.4027974	total: 96.1ms	remaining: 1.5s
6:	learn: 1.3455860	total: 111ms	remaining: 1.47s
7:	learn: 1.2961213	total: 126ms	remaining: 1.45s
8:	learn: 1.2531179	total: 141ms	remaining: 1.43s
9:	learn: 1.2172486	total: 157ms	remaining: 1.41s
10:	learn: 1.1856831	total: 171ms	remaining: 1.38s
11:	learn: 1.1589271	total: 180ms	remaining: 1.32s
12:	learn: 1.1360263	total: 195ms	remaining: 1.3s
13:	learn: 1.1160686	total: 211ms	remaining: 1.3s
14:	learn: 1.0942570	total: 226ms	remaining: 1.28s
15:	learn: 1.0776286	total: 241ms	remaining: 1.26s
16:	learn: 1.0597891	total: 257ms	remaining: 1.25s
17:	learn: 1.0425900	total: 273ms	remaining: 1.24s
18:	learn: 1.0287235	total: 288ms	remaining: 1.23s
19:	learn: 1.0176531	total: 302ms	remai

In [35]:
cat_pred = cat_model.predict(poker_testip)

In [36]:
report = classification_report(cat_pred, poker_testop)
confusion = confusion_matrix(cat_pred, poker_testop)
print("Classification Report:\n", report)
print("Confusion Matrix:\n", confusion)

C:\Users\arvin\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\metrics\_classification.py:1469: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\arvin\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\metrics\_classification.py:1469: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


Classification Report:
               precision    recall  f1-score   support

           0       0.83      0.65      0.73    642089
           1       0.51      0.60      0.55    356845
           2       0.01      0.60      0.02       665
           3       0.02      0.89      0.03       401
           4       0.00      0.00      0.00         0
           5       0.00      0.00      0.00         0
           6       0.00      0.00      0.00         0
           7       0.00      0.00      0.00         0
           8       0.00      0.00      0.00         0
           9       0.00      0.00      0.00         0

    accuracy                           0.63   1000000
   macro avg       0.14      0.27      0.13   1000000
weighted avg       0.71      0.63      0.66   1000000

Confusion Matrix:
 [[416168 208072  11520   4247    194   1827     53      5      0      3]
 [ 85041 214302  35700  16451   3691    169   1276    203     12      0]
 [     0    123    402     65      0      0     60  

C:\Users\arvin\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\metrics\_classification.py:1469: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


Light Boost

In [37]:
light_model = LGBMClassifier(learning_rate=0.1, n_estimators=100, max_depth=6, objective='multiclass')


In [38]:
light_model.fit (poker_ip, poker_op)

C:\Users\arvin\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\preprocessing\_label.py:97: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\arvin\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\preprocessing\_label.py:132: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, dtype=self.classes_.dtype, warn=True)


[LightGBM] [Warning] Accuracy may be bad since you didn't explicitly set num_leaves OR 2^max_depth > num_leaves. (num_leaves=31).
[LightGBM] [Warning] Found whitespace in feature_names, replace with underlines
[LightGBM] [Warning] Accuracy may be bad since you didn't explicitly set num_leaves OR 2^max_depth > num_leaves. (num_leaves=31).
[LightGBM] [Warning] Auto-choosing row-wise multi-threading, the overhead of testing was 0.000307 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 95
[LightGBM] [Info] Number of data points in the train set: 25010, number of used features: 10
[LightGBM] [Info] Start training from score -0.694107
[LightGBM] [Info] Start training from score -0.858516
[LightGBM] [Info] Start training from score -3.031967
[LightGBM] [Info] Start training from score -3.886755
[LightGBM] [Info] Start training from score -5.594432
[LightGBM] [Info] Start training fro

LGBMClassifier(max_depth=6, objective='multiclass')

In [39]:
light_pred = light_model.predict(poker_testip)

[LightGBM] [Warning] Accuracy may be bad since you didn't explicitly set num_leaves OR 2^max_depth > num_leaves. (num_leaves=31).


In [40]:
report = classification_report(light_pred, poker_testop)
confusion = confusion_matrix(light_pred, poker_testop)
print("Classification Report:\n", report)
print("Confusion Matrix:\n", confusion)

Classification Report:
               precision    recall  f1-score   support

           0       0.77      0.65      0.70    588184
           1       0.51      0.58      0.54    370593
           2       0.04      0.20      0.07      9493
           3       0.05      0.20      0.08      4897
           4       0.03      0.03      0.03      3342
           5       0.00      0.00      0.00     10176
           6       0.01      0.00      0.01      4699
           7       0.01      0.00      0.00      4128
           8       0.08      0.00      0.00      3414
           9       0.00      0.00      0.00      1074

    accuracy                           0.60   1000000
   macro avg       0.15      0.17      0.14   1000000
weighted avg       0.64      0.60      0.62   1000000

Confusion Matrix:
 [[383674 189202   9250   3952    438   1575     81     10      1      1]
 [102223 214222  34588  14876   3201    265   1060    148      8      2]
 [  1582   5201   1878    620     22      4    166  

Now performing Oversampling

In [41]:
poker_op.value_counts()

Poker Hand
0             12493
1             10599
2              1206
3               513
4                93
5                54
6                36
7                 6
8                 5
9                 5
Name: count, dtype: int64

In [42]:
from imblearn.over_sampling import SMOTE

In [43]:
smote = SMOTE(sampling_strategy='minority', k_neighbors=4)

In [44]:
X_smt, y_smt = smote.fit_resample(poker_ip, poker_op)

In [45]:
light_model.fit (X_smt, y_smt)

[LightGBM] [Warning] Accuracy may be bad since you didn't explicitly set num_leaves OR 2^max_depth > num_leaves. (num_leaves=31).
[LightGBM] [Warning] Found whitespace in feature_names, replace with underlines
[LightGBM] [Warning] Accuracy may be bad since you didn't explicitly set num_leaves OR 2^max_depth > num_leaves. (num_leaves=31).
[LightGBM] [Warning] Auto-choosing row-wise multi-threading, the overhead of testing was 0.000541 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 95
[LightGBM] [Info] Number of data points in the train set: 37498, number of used features: 10
[LightGBM] [Info] Start training from score -1.099119
[LightGBM] [Info] Start training from score -1.263528
[LightGBM] [Info] Start training from score -3.436978
[LightGBM] [Info] Start training from score -4.291767
[LightGBM] [Info] Start training from score -5.999443
[LightGBM] [Info] Start training fro

C:\Users\arvin\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\preprocessing\_label.py:97: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\arvin\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\preprocessing\_label.py:132: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, dtype=self.classes_.dtype, warn=True)


[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No f

LGBMClassifier(max_depth=6, objective='multiclass')

In [46]:
light_pred = light_model.predict(poker_testip)

[LightGBM] [Warning] Accuracy may be bad since you didn't explicitly set num_leaves OR 2^max_depth > num_leaves. (num_leaves=31).


In [47]:
report = classification_report(light_pred, poker_testop)
confusion = confusion_matrix(light_pred, poker_testop)
print("Classification Report:\n", report)
print("Confusion Matrix:\n", confusion)

Classification Report:
               precision    recall  f1-score   support

           0       0.80      0.69      0.74    583882
           1       0.57      0.61      0.59    392553
           2       0.04      0.30      0.07      6355
           3       0.05      0.36      0.08      2725
           4       0.02      0.08      0.03       822
           5       0.18      0.11      0.13      3420
           6       0.00      0.01      0.01       917
           7       0.01      0.00      0.00      4557
           8       0.17      0.00      0.00      4263
           9       0.00      0.00      0.00       506

    accuracy                           0.64   1000000
   macro avg       0.18      0.22      0.16   1000000
weighted avg       0.69      0.64      0.66   1000000

Confusion Matrix:
 [[401289 172996   6074   2015    102   1397      7      1      1      0]
 [ 93532 238716  38186  17189   3504    184   1089    147      5      1]
 [   442   3336   1910    463     15      0    172  

In [48]:
cat_model.fit (X_smt, y_smt, cat_features=cat_features)

0:	learn: 1.9099048	total: 30.6ms	remaining: 3.03s
1:	learn: 1.6895729	total: 58.8ms	remaining: 2.88s
2:	learn: 1.5133247	total: 85.8ms	remaining: 2.77s
3:	learn: 1.3832167	total: 112ms	remaining: 2.68s
4:	learn: 1.2823047	total: 137ms	remaining: 2.6s
5:	learn: 1.1978242	total: 161ms	remaining: 2.51s
6:	learn: 1.1294876	total: 188ms	remaining: 2.5s
7:	learn: 1.0721576	total: 212ms	remaining: 2.44s
8:	learn: 1.0239762	total: 237ms	remaining: 2.4s
9:	learn: 0.9830656	total: 261ms	remaining: 2.35s
10:	learn: 0.9472849	total: 286ms	remaining: 2.31s
11:	learn: 0.9168284	total: 309ms	remaining: 2.27s
12:	learn: 0.8904050	total: 332ms	remaining: 2.22s
13:	learn: 0.8645141	total: 353ms	remaining: 2.17s
14:	learn: 0.8435701	total: 377ms	remaining: 2.13s
15:	learn: 0.8228633	total: 399ms	remaining: 2.1s
16:	learn: 0.8030751	total: 424ms	remaining: 2.07s
17:	learn: 0.7883997	total: 448ms	remaining: 2.04s
18:	learn: 0.7732908	total: 472ms	remaining: 2.01s
19:	learn: 0.7594444	total: 497ms	remainin

In [49]:
cat_pred = cat_model.predict(poker_testip)

In [50]:
report = classification_report(cat_pred, poker_testop)
confusion = confusion_matrix(cat_pred, poker_testop)
print("Classification Report:\n", report)
print("Confusion Matrix:\n", confusion)

C:\Users\arvin\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\metrics\_classification.py:1469: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\arvin\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\metrics\_classification.py:1469: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


Classification Report:
               precision    recall  f1-score   support

           0       0.82      0.64      0.72    639511
           1       0.50      0.59      0.54    356795
           2       0.01      0.63      0.01       558
           3       0.02      0.85      0.03       385
           4       0.00      0.00      0.00         0
           5       0.00      0.00      0.00         0
           6       0.00      0.00      0.00         0
           7       0.00      0.00      0.00         0
           8       0.25      0.00      0.00      2751
           9       0.00      0.00      0.00         0

    accuracy                           0.62   1000000
   macro avg       0.16      0.27      0.13   1000000
weighted avg       0.70      0.62      0.65   1000000

Confusion Matrix:
 [[410852 209957  11878   4696    241   1800     77      7      0      3]
 [ 89517 211012  35141  15996   3531    112   1270    207      9      0]
 [     0    130    351     36      0      0     32  

C:\Users\arvin\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\metrics\_classification.py:1469: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


Conclusion

Without Sampling-
CatBoost- F1=81%, Acc=73%
LightBoost- f1=70%, Acc=60%

With SMOTE-
CatBoost- F1=79%, Acc=70%
LightBoost- f1=67%, Acc=57%